libclc/ptx-nvidiacl/libspirv/group/group_ballot.cl

```diff
@@ -0,0 +1,35 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include <spirv/spirv.h>
+#include <spirv/spirv_types.h>
+
+_CLC_DEF
+__clc_vec4_uint32_t _Z29__spirv_GroupNonUniformBallotjb(unsigned flag,
+                                                        bool predicate) {
+  // only support subgroup for now
+  if (flag != Subgroup) {
+    __builtin_trap();
+    __builtin_unreachable();
+  }
+
+  // prepare result, we only support the ballot operation on 32 threads maximum
+  // so we only need the first element to represent the final mask
+  __clc_vec4_uint32_t res;
+  res[1] = 0;
+  res[2] = 0;
+  res[3] = 0;
+
+  // evaluate the predicate on all threads
+  unsigned threads = 0xFFFFFFFF;
```

Comment:
Is this always safe?  What happens if the warp is the last warp in the block and things don't divide evenly?

There are other parts of libspirv where we're defensively computing masks inside every call, and we know this is degrading performance (see https://github.com/intel/llvm/issues/4504).  If using `0xFFFFFFFF` is guaranteed to work, we should adopt this style in the other libspirv functions as an optimization.

Reply:
That's a very good question, and I'm not so sure anymore, originally I thought it was safe because of the following wording in the PTX ISA for `vote.sync.ballot`:

> vote.sync will cause executing thread to wait until all non-exited threads corresponding to membermask have executed vote.sync with the same qualifiers and same membermask value before resuming execution.

If I'm reading this correctly it means the instruction will wait on an intersection of "non-exited" threads and threads specified in the "membermask", if that is the case, then I would hope that in a warp smaller than 32 threads, the unused threads would be considered as "exited" and safely skipped.

The `SYCL/SubGroupMask/GroupSize.cpp` test is actually testing this with different subgroup sizes `8`, 16`, `32`, which I hope would turn into "incomplete" warps and exercise this, and this test passes fine with this implementation for me.

However going through #4504 I can see where you're coming from, this is a little ambiguous, the NVVM IR docs and PTX ISA don't have quite the same wording, the NVVM IR for `vote.sync.ballot` has:

> The behavior of this intrinsic is undefined if any thread participating in the intrinsic has exited or the executing thread is not in the %membermask.

Interestingly that specific line is not there in the PTX ISA.

Anyway, I agree with you that this might be unsafe, I'll go with the defensive mask computation in this PR for now, and move back to #4504 for further discussion.

Reply:
Updated to use `membermask`